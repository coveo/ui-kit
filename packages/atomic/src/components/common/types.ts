export type SearchStatus = {
  state: {
    hasResults: boolean;
  };
};

export type Pager = {
  isCurrentPage(page: number): boolean;
  selectPage(page: number): void;
  previousPage(): void;
  nextPage(): void;
  state: {
    hasNextPage: boolean;
    hasPreviousPage: boolean;
    currentPages: number[];
  };
};
