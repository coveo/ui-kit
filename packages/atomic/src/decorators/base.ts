// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type Constructor<T = {}> = new (...args: any[]) => T;
