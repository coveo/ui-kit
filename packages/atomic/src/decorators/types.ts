import {TemplateResult} from 'lit';
import {TemplateResultType} from 'lit/directive-helpers.js';
import {AnyBindings} from '../components/common/interface/bindings';

export type GenericRender<T extends TemplateResultType> = TemplateResult<T>;

export type RenderGuardDecorator<
  Component,
  T extends TemplateResultType,
  Descriptor = TypedPropertyDescriptor<() => GenericRender<T>>,
> = (
  target: Component,
  propertyKey: 'render',
  descriptor: Descriptor
) => Descriptor;

/**
 * Necessary interface an Atomic Component must have to initialize itself correctly.
 */
export interface InitializableComponent<
  SpecificBindings extends AnyBindings = AnyBindings,
> {
  /**
   * Bindings passed from the `AtomicSearchInterface` to its children components.
   */
  bindings: SpecificBindings;

  /**
   * Flag to check if the component has been initialized.
   */
  initialized?: boolean;

  /**
   * Method called when the component is removed from the DOM.
   */
  unsubscribeLanguage?: () => void;

  /**
   * Method called right after the `bindings` property is defined. This is the method where Headless Framework controllers should be initialized.
   */
  initialize?: () => void;
  error: Error;
}
