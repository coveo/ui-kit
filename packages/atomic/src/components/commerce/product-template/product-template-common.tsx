import {
  ProductTemplate,
  ProductTemplateCondition,
  ProductTemplatesHelpers,
} from '@coveo/headless/commerce';
import {h} from '@stencil/core';
import {aggregate, isElementNode, isVisualNode} from '../../../utils/utils';
import {isResultSectionNode} from '../../common/layout/sections';
import {tableElementTagName} from '../../search/atomic-table-result/table-element-utils';

export type TemplateContent = DocumentFragment;

interface ProductTemplateCommonProps {
  allowEmpty?: boolean;
  host: HTMLDivElement;
  validParents: string[];
  setError: (error: Error) => void;
}

type TemplateNodeType =
  | 'section'
  | 'metadata'
  | 'table-column-definition'
  | 'other';

export function getTemplateNodeType(node: Node): TemplateNodeType {
  if (isResultSectionNode(node)) {
    return 'section';
  }
  if (!isVisualNode(node)) {
    return 'metadata';
  }
  if (
    isElementNode(node) &&
    node.tagName.toLowerCase() === tableElementTagName
  ) {
    return 'table-column-definition';
  }
  return 'other';
}

function groupNodesByType(nodes: NodeList) {
  return aggregate(Array.from(nodes), (node) => getTemplateNodeType(node));
}

export class ProductTemplateCommon {
  private host: HTMLDivElement;
  public matchConditions: ProductTemplateCondition[] = [];

  constructor({
    host,
    setError,
    validParents,
    allowEmpty = false,
  }: ProductTemplateCommonProps) {
    this.host = host;
    this.validateTemplate(host, setError, validParents, allowEmpty);
  }

  validateTemplate(
    host: HTMLDivElement,
    setError: (error: Error) => void,
    validParents: string[],
    allowEmpty = true
  ) {
    const hasValidParent = validParents
      .map((p) => p.toUpperCase())
      .includes(host.parentElement?.nodeName || '');
    const tagName = host.nodeName.toLowerCase();

    if (!hasValidParent) {
      setError(
        new Error(
          `The "${tagName}" component has to be the child of one of the following: ${validParents
            .map((p) => `"${p.toLowerCase()}"`)
            .join(', ')}.`
        )
      );
      return;
    }

    const template = host.querySelector('template');
    if (!template) {
      setError(
        new Error(
          `The "${tagName}" component must contain a "template" element as a child.`
        )
      );
      return;
    }

    if (!allowEmpty && !template.innerHTML.trim()) {
      setError(
        new Error(`The "template" tag inside "${tagName}" cannot be empty.`)
      );
      return;
    }

    if (template.content.querySelector('script')) {
      console.warn(
        'Any "script" tags defined inside of "template" elements are not supported and will not be executed when the products are rendered.',
        host
      );
    }

    const {section: sectionNodes, other: otherNodes} = groupNodesByType(
      template.content.childNodes
    );
    if (sectionNodes?.length && otherNodes?.length) {
      console.warn(
        'Product templates should only contain section elements or non-section elements. Future updates could unpredictably affect this product template.',
        host,
        {sectionNodes, otherNodes}
      );
    }
  }

  getTemplate(
    conditions: ProductTemplateCondition[],
    error: Error
  ): ProductTemplate<TemplateContent> | null {
    if (error) {
      return null;
    }

    return {
      conditions: conditions.concat(this.matchConditions),
      content: getTemplateElement(this.host).content!,
      priority: 1,
    };
  }

  renderIfError(error: Error) {
    if (error) {
      return (
        <atomic-component-error
          element={this.host}
          error={error}
        ></atomic-component-error>
      );
    }
  }
}

function getTemplateElement(host: HTMLElement) {
  return host.querySelector('template')!;
}
export function makeMatchConditions(
  mustMatch: Record<string, string[]>,
  mustNotMatch: Record<string, string[]>
): ProductTemplateCondition[] {
  const conditions: ProductTemplateCondition[] = [];
  for (const field in mustMatch) {
    conditions.push(
      ProductTemplatesHelpers.fieldMustMatch(field, mustMatch[field])
    );
  }

  for (const field in mustNotMatch) {
    conditions.push(
      ProductTemplatesHelpers.fieldMustNotMatch(field, mustNotMatch[field])
    );
  }
  return conditions;
}

export function makeDefinedConditions(
  ifDefined?: string,
  ifNotDefined?: string
): ProductTemplateCondition[] {
  const conditions: ProductTemplateCondition[] = [];
  if (ifDefined) {
    const fieldNames = ifDefined.split(',');
    conditions.push(ProductTemplatesHelpers.fieldsMustBeDefined(fieldNames));
  }

  if (ifNotDefined) {
    const fieldNames = ifNotDefined.split(',');
    conditions.push(ProductTemplatesHelpers.fieldsMustNotBeDefined(fieldNames));
  }
  return conditions;
}
