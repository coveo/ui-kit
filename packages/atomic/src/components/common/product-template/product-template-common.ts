import {
  ProductTemplateCondition,
  ProductTemplatesHelpers,
} from '@coveo/headless/commerce';
import {intersection} from '../../../utils/set';
import {isElementNode, isVisualNode} from '../../../utils/utils';
import {tableElementTagName} from '../../search/atomic-table-result/table-element-utils';
import {isResultSectionNode} from '../layout/sections';

type TemplateNodeType =
  | 'section'
  | 'metadata'
  | 'table-column-definition'
  | 'other';

export function getTemplateNodeType(node: Node): TemplateNodeType {
  if (isResultSectionNode(node)) {
    return 'section';
  }
  if (!isVisualNode(node)) {
    return 'metadata';
  }
  if (
    isElementNode(node) &&
    node.tagName.toLowerCase() === tableElementTagName
  ) {
    return 'table-column-definition';
  }
  return 'other';
}

export function makeMatchConditions(
  mustMatch: Record<string, string[]>,
  mustNotMatch: Record<string, string[]>
): ProductTemplateCondition[] {
  const conditions: ProductTemplateCondition[] = [];
  for (const field in mustMatch) {
    if (mustNotMatch[field]) {
      const mustNotMatchValues = new Set(mustNotMatch[field]);
      const mustMatchValues = new Set(mustMatch[field]);
      const commonValues = intersection(mustNotMatchValues, mustMatchValues);
      if (commonValues.size > 0) {
        console.error(
          `Conflicting match conditions for field ${field}, the template will be ignored.`,
          commonValues
        );
        return [() => false];
      }
    }
    conditions.push(
      ProductTemplatesHelpers.fieldMustMatch(field, mustMatch[field])
    );
  }

  for (const field in mustNotMatch) {
    conditions.push(
      ProductTemplatesHelpers.fieldMustNotMatch(field, mustNotMatch[field])
    );
  }
  return conditions;
}

export function makeDefinedConditions(
  ifDefined?: string,
  ifNotDefined?: string
): ProductTemplateCondition[] {
  const conditions: ProductTemplateCondition[] = [];
  if (ifDefined) {
    const fieldNames = ifDefined.split(',');
    conditions.push(ProductTemplatesHelpers.fieldsMustBeDefined(fieldNames));
  }

  if (ifNotDefined) {
    const fieldNames = ifNotDefined.split(',');
    conditions.push(ProductTemplatesHelpers.fieldsMustNotBeDefined(fieldNames));
  }
  return conditions;
}
