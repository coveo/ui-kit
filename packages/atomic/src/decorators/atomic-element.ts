import {Constructor} from './base';

/**
 * Allow for custom element classes with private constructors
 */
type CustomElementClass = Omit<typeof HTMLElement, 'new'>;

export type CustomElementDecorator = {
  (target: CustomElementClass): void;
};

/**
 * Class decorator factory that defines the decorated class as a custom element.
 *
 * Replaces the `@customElement` decorator from Lit with a implementation that
 * checks if the custom element is already defined before defining it.
 *
 * ```js
 * @atomicElement('my-element')
 * class MyElement extends LitElement {
 *   render() {
 *     return html``;
 *   }
 * }
 * ```
 * @param tagName The tag name of the atomic element to define.
 */
export const atomicElement =
  (tagName: string): CustomElementDecorator =>
  (classOrTarget: CustomElementClass | Constructor<HTMLElement>) => {
    if (!customElements.get(tagName)) {
      customElements.define(tagName, classOrTarget as CustomElementConstructor);
    }
  };
