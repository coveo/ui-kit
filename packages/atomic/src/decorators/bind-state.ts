import type {Controller} from '@coveo/headless';
import type {ReactiveElement} from 'lit';
import type {InitializableComponent} from './types';

type ControllerProperties<T> = {
  [K in keyof T]: T[K] extends Controller ? K : never;
}[keyof T];

/**
 * A decorator that allows the Lit component state property to automatically get updates from a [Coveo Headless controller](https://docs.coveo.com/en/headless/latest/usage/#use-headless-controllers).
 *
 * @example
 * ```ts
 * @bindStateToController('pager') @state() private pagerState!: PagerState;
 * ```
 *
 * For more information and examples, view the "Utilities" section of the readme.
 *
 * @param controllerProperty The controller property to subscribe to. The controller has to be created inside of the `initialize` method.
 * @param options The configurable `bindStateToController` options.
 */
export function bindStateToController<Element extends ReactiveElement>(
  controllerProperty: ControllerProperties<Element>,
  options?: {
    /**
     * Component's method to be called when state is updated.
     */
    onUpdateCallbackMethod?: string;
  }
) {
  return <
    T extends Record<ControllerProperties<Element>, Controller> &
      Record<string, unknown>,
    Instance extends Element & T & InitializableComponent,
    K extends keyof Instance,
  >(
    proto: Element,
    stateProperty: K
  ) => {
    const ctor = proto.constructor as typeof ReactiveElement;

    ctor.addInitializer((instance) => {
      const component = instance as Instance;
      const {disconnectedCallback, initialize} = component;

      component.initialize = function () {
        initialize?.call(this);

        if (!initialize) {
          return console.error(
            `ControllerState: The "initialize" method has to be defined and instantiate a controller for the property ${controllerProperty.toString()}`,
            component
          );
        }

        if (!component[controllerProperty]) {
          return console.error(
            `${controllerProperty.toString()} property is not defined on component`,
            component
          );
        }

        if (
          options?.onUpdateCallbackMethod &&
          !component[options.onUpdateCallbackMethod]
        ) {
          return console.error(
            `ControllerState: The onUpdateCallbackMethod property "${options.onUpdateCallbackMethod}" is not defined`,
            component
          );
        }

        const controller = component[controllerProperty];
        const updateCallback = options?.onUpdateCallbackMethod
          ? component[options.onUpdateCallbackMethod]
          : undefined;

        const unsubscribeController = controller.subscribe(() => {
          component[stateProperty] = controller.state as Instance[K];
          typeof updateCallback === 'function' && updateCallback();
        });

        component.disconnectedCallback = () => {
          !component.isConnected && unsubscribeController?.();
          disconnectedCallback?.call(component);
        };
      };
    });
  };
}
