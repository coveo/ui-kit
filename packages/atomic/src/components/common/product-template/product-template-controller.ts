import type {
  ProductTemplate,
  ProductTemplateCondition,
} from '@coveo/headless/commerce';
import type {ReactiveController, ReactiveControllerHost} from 'lit';
import {aggregate} from '../../../utils/utils';
import type {ItemTarget} from '../../common/layout/display-options';
import {getTemplateNodeType} from './product-template-common';

export type TemplateContent = DocumentFragment;

type ProductTemplateHost = ReactiveControllerHost &
  HTMLElement & {error: Error};

export class ProductTemplateController implements ReactiveController {
  public matchConditions: ProductTemplateCondition[] = [];
  private gridCellLinkTarget?: ItemTarget;

  private static readonly ERRORS = {
    invalidParent: (tagName: string, validParents: string[]) =>
      `The "${tagName}" component has to be the child of one of the following: ${validParents
        .map((p) => `"${p.toLowerCase()}"`)
        .join(', ')}.`,
    missingTemplate: (tagName: string) =>
      `The "${tagName}" component must contain a "template" element as a child.`,
    emptyTemplate: (tagName: string) =>
      `The "template" tag inside "${tagName}" cannot be empty.`,
  };

  private static readonly WARNINGS = {
    scriptTag:
      'Any "script" tags defined inside of "template" elements are not supported and will not be executed when the products are rendered.',
    sectionMix:
      'Product templates should only contain section elements or non-section elements. Future updates could unpredictably affect this product template.',
  };

  constructor(
    private host: ProductTemplateHost,
    private validParents: string[],
    private allowEmpty: boolean = false
  ) {
    this.host.addController(this);
  }

  hostConnected() {
    this.validateTemplate();
  }

  setError(error: Error) {
    this.host.error = error;
  }

  validateTemplate() {
    const hasValidParent = this.validParents
      .map((p) => p.toUpperCase())
      .includes(this.parentElement?.nodeName || '');
    const tagName = this.host.nodeName.toLowerCase();

    if (!hasValidParent) {
      this.setError(
        new Error(
          ProductTemplateController.ERRORS.invalidParent(
            tagName,
            this.validParents
          )
        )
      );
      return;
    }

    if (this.parentAttr('display') === 'grid') {
      this.gridCellLinkTarget = this.parentAttr(
        'grid-cell-link-target'
      ) as ItemTarget;
    }

    if (!this.template) {
      this.setError(
        new Error(ProductTemplateController.ERRORS.missingTemplate(tagName))
      );
      return;
    }

    if (!this.allowEmpty && !this.template.innerHTML.trim()) {
      this.setError(
        new Error(ProductTemplateController.ERRORS.emptyTemplate(tagName))
      );
      return;
    }

    if (this.template.content.querySelector('script')) {
      console.warn(ProductTemplateController.WARNINGS.scriptTag, this.host);
    }

    const {section, other} = groupNodesByType(this.template.content.childNodes);

    if (section?.length && other?.length) {
      console.warn(ProductTemplateController.WARNINGS.sectionMix, this.host, {
        section,
        other,
      });
    }
  }

  getTemplate(
    conditions: ProductTemplateCondition[]
  ): ProductTemplate<TemplateContent> | null {
    if (this.host.error) {
      return null;
    }
    return {
      conditions: conditions.concat(this.matchConditions),
      content: getTemplateElement(this.host).content!,
      linkContent: this.getLinkTemplateElement(this.host).content!,
      priority: 1,
    };
  }

  getDefaultLinkTemplateElement() {
    const linkTemplate = document.createElement('template');
    linkTemplate.innerHTML = `<atomic-product-link>${this.gridCellLinkTarget ? `<a slot="attributes" target="${this.gridCellLinkTarget}"></a>` : ''}</atomic-product-link>`;
    return linkTemplate;
  }

  getLinkTemplateElement(host: HTMLElement) {
    return (
      host.querySelector<HTMLTemplateElement>('template[slot="link"]') ??
      this.getDefaultLinkTemplateElement()
    );
  }

  private get parentElement() {
    return this.host.parentElement;
  }

  private get template() {
    return this.host.querySelector('template');
  }

  private parentAttr(attribute: string) {
    return this.parentElement?.attributes.getNamedItem(attribute)?.value;
  }
}

function getTemplateElement(host: HTMLElement) {
  return host.querySelector<HTMLTemplateElement>('template:not([slot])')!;
}

function groupNodesByType(nodes: NodeList) {
  return aggregate(Array.from(nodes), (node) => getTemplateNodeType(node));
}
