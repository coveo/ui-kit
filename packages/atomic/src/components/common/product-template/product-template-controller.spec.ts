import {LitElementWithError} from '@/src/decorators/types';
import {fixture} from '@/vitest-utils/testing-helpers/fixture';
import {html, LitElement, TemplateResult} from 'lit';
import {customElement, state} from 'lit/decorators.js';
import {describe, it, expect, vi} from 'vitest';
import {getTemplateNodeType} from './product-template-common';
import {ProductTemplateController} from './product-template-controller';

vi.mock('./product-template-common', {spy: true});

@customElement('test-element')
class TestElement extends LitElement implements LitElementWithError {
  @state()
  public error!: Error;
  controller = new ProductTemplateController(this, ['valid-parent'], false);
}

@customElement('empty-test-element')
class EmptyTestElement extends LitElement implements LitElementWithError {
  @state()
  public error!: Error;
  controller = new ProductTemplateController(this, ['valid-parent'], true);
}

// @customElement('valid-parent')
// class ValidParent extends LitElement {}

// @customElement('invalid-parent')
// class InvalidParent extends LitElement {}

describe('ProductTemplateController', () => {
  async function setupElement(
    template: TemplateResult<1>,
    parentNode: HTMLElement = document.createElement('valid-parent')
  ) {
    await fixture(template, parentNode);
    return document.querySelector('test-element')! as TestElement;
  }

  describe('when the host has not a valid parent', () => {
    it('should set an error', async () => {
      const element = await setupElement(
        html`<test-element>
          <template><h1>hello</h1></template>
        </test-element>`,
        document.createElement('invalid-parent')
      );

      expect(element.error).toBeInstanceOf(Error);
      expect(element.error.message).toContain('has to be the child');
    });
  });

  describe('when the template is missing from the host', () => {
    it('should set an error', async () => {
      const element = await setupElement(html`<test-element></test-element>`);

      expect(element.error).toBeInstanceOf(Error);
      expect(element.error.message).toContain('must contain a "template"');
    });
  });

  describe('when the template is empty', () => {
    it('should set an error if allowEmpty is false', async () => {
      const element = await setupElement(
        html`<test-element>
          <template> </template>
        </test-element>`
      );

      expect(element.error).toBeInstanceOf(Error);
      expect(element.error.message).toContain('cannot be empty');
    });

    it('should not set an error if allowEmpty is true', async () => {
      await setupElement(
        html`<empty-test-element>
          <template> </template>
        </empty-test-element>`
      );

      const element = document.querySelector(
        'empty-test-element'
      ) as EmptyTestElement;

      expect(element.error).toBeUndefined();
    });
  });

  describe('when the template contains script tags', () => {
    it('should log a warning', async () => {
      const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
      const element = await setupElement(
        html`<test-element>
          <template><script></script></template>
        </test-element>`
      );

      expect(warnSpy).toHaveBeenCalledWith(
        expect.stringContaining('script'),
        element
      );

      warnSpy.mockRestore();
    });
  });

  describe('when the template contains both section and other nodes', () => {
    const localSetup = () =>
      setupElement(
        html`<test-element>
          <template>
            <atomic-result-section-visual>section</atomic-result-section-visual>
            <span>other</span>
          </template>
        </test-element>`
      );

    it('should call #getTemplateNodeType with the appropriate sections', async () => {
      const mockedGetTemplateNodeType = vi.mocked(getTemplateNodeType);
      await localSetup();

      expect(mockedGetTemplateNodeType.mock.calls).toMatchSnapshot();
    });

    it('should log a warning', async () => {
      const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
      await localSetup();

      expect(warnSpy).toHaveBeenCalledWith(
        expect.stringContaining('should only contain section elements'),
        expect.any(TestElement),
        expect.objectContaining({})
      );
    });
  });

  //   describe.skip('when the template is valid', () => {});

  //   beforeEach(async () => {
  //     await fixture(html`<test-element></test-element>`);
  //   });

  //   it('sets error if missing template', () => {
  //     host.innerHTML = '';
  //     host.controller.validateTemplate();
  //     expect(host.controller['error']).toBeInstanceOf(Error);
  //     expect(host.controller['error']!.message).toContain(
  //       'must contain a "template"'
  //     );
  //   });

  //   it('sets error if template is empty and allowEmpty is false', () => {
  //     host.innerHTML = '<template>   </template>';
  //     host.controller.validateTemplate();
  //     expect(host.controller['error']).toBeInstanceOf(Error);
  //     expect(host.controller['error']!.message).toContain('cannot be empty');
  //   });

  //   it('does not set error if template is empty and allowEmpty is true', () => {
  //     ({host} = createHostWithTemplate({
  //       template: '<template>   </template>',
  //       allowEmpty: true,
  //     }));
  //     host.controller.validateTemplate();
  //     expect(host.controller['error']).toBeNull();
  //   });

  //   it('warns if template contains script', () => {
  //     const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
  //     host.innerHTML = '<template><script></script></template>';
  //     host.controller.validateTemplate();
  //     expect(warnSpy).toHaveBeenCalledWith(
  //       expect.stringContaining('script'),
  //       host
  //     );
  //     warnSpy.mockRestore();
  //   });

  //   it('warns if template contains both section and other nodes', () => {
  //     const warnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});
  //     // Mock getTemplateNodeType to alternate between 'section' and 'other'
  //     vi.mock('./product-template-common', () => ({
  //       getTemplateNodeType: (node: Node) =>
  //         node.textContent === 'section' ? 'section' : 'other',
  //     }));
  //     host.innerHTML =
  //       '<template><div>section</div><span>other</span></template>';
  //     host.controller.validateTemplate();
  //     expect(warnSpy).toHaveBeenCalledWith(
  //       expect.stringContaining('should only contain section elements'),
  //       host,
  //       expect.objectContaining({
  //         sectionNodes: expect.any(Array),
  //         otherNodes: expect.any(Array),
  //       })
  //     );
  //     warnSpy.mockRestore();
  //   });

  //   it('getTemplate returns null if error is set', () => {
  //     host.controller['error'] = new Error('fail');
  //     const result = host.controller.getTemplate([]);
  //     expect(result).toBeNull();
  //   });

  //   it('getTemplate returns template object if no error', () => {
  //     const result = host.controller.getTemplate([]);
  //     expect(result).not.toBeNull();
  //     expect(result).toHaveProperty('conditions');
  //     expect(result).toHaveProperty('content');
  //     expect(result).toHaveProperty('linkContent');
  //     expect(result).toHaveProperty('priority');
  //   });
});
