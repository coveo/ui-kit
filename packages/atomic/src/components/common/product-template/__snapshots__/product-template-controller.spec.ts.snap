// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductTemplateController > when the template contains both section and other nodes > should call #getTemplateNodeType with the appropriate sections 1`] = `
[
  [
    
            ,
  ],
  [
    <atomic-result-section-visual>
      section
    </atomic-result-section-visual>,
  ],
  [
    
            ,
  ],
  [
    <span>
      other
    </span>,
  ],
  [
    
          ,
  ],
]
`;
